### ISSUE TITLE
Document and Research - How to make new brush copies with new settings and icons in the toolshelf

### ISSUE THREAD

**Issue by Draise14** [Bforartists/Bforartists#2164]
# Research
![image](https://user-images.githubusercontent.com/35357409/104213075-ecd5c600-5403-11eb-97f3-ba2f7d53d5d9.png)
This task is to document and research how to create brush copies with new icons and new brush settings to expose default brushes to top level. 

The reason being is that there are many brushes in the paint, weight, vertex paint and other modes that has a single toolshelf entry but 7+ brush settings that have to be accessed through a dropdown. This is a very slow workflow, and completely unnecessary. 

We need to know how hard this would be to code, if it's worth maintaining this code before Blender eventually updates this with a yet unworked syste, and if this is even possible. 

A related feature request:
#2136

# Task
Find a way how to do this easily. 
- Create a mockup of one extra brush with new settings, with icon
- Check and see if each brush needs a new library of sub-brushes per brush, or if they are all linked to the same brush settings library

Once the mockup is done, this closes and new sub-tasks are made per tool on a per mode per editor per brush setting basis. 

# Sub-Task
This may need to find a way to generate icons for the toolshelf from the *.SVG/*.PNG first, as currently the toolshelf icons are generated from 3D objects in a blend. 

Priority could be this task:
#1318 

## Method
I've asked @iyadahmed to do some research into this, document it and find away to save some time figuring it out to give us some options before the Blender devs update their way. They have had slow progress to update the brush system, but if this is easy to do, the extra brushes could be more accessible.

**Comment by ReinerBforartists** [comment]
Draise, this is under development by the Blender devs. They want to redo the whole brush system at one point. It is no good idea when we start to do the same job now.

**Comment by Draise14** [comment]
Relevant task is this one... which is still in UX proposal stage. There are no actual commits, branch or tasks for the redesign yet, so I doubt it will be out till post Blender 3.1.0+ days. 
[Blender Developer Task](https://developer.blender.org/T56744)

If this is easy, we can quickly implement it and benefit from the UX instead of waiting more than a year or more for the new system.

**Comment by ReinerBforartists** [comment]
It is not easy. This task means to redo all brushes and the whole brush system. That's why i am not really happy to do it.

**Comment by Draise14** [comment]
As mentioned, this 2.92 roadmap is still a drawing board when it comes to brushes, and 2.92 is now Bcon3, soon another release... No code, nearing an end to a dev cycle, it's definitely not coming out for 2.92. 

Just to clarify, you don't need to take a look. This is also for @iyadahmed to take a look to see - no sweat. This task will be closed if we see if the code is easy or hard/impossible to duplicate a brush with a new setting from the same settings library or each their own settings library. 

Motive: I've been needing these this very month with work so I asked him to take a look at it, and made a task for it to keep things open - because burning hours going through a drop down to toggle a brush, and sometimes switching a brush AND toggling its settings is driving me crazy - and their design document and proposal for the redesign fixes nothing really. So I am sponsoring the research and labour for this experiment.

**Comment by ReinerBforartists** [comment]
It is wasted time. And it is a HUGE task. And we pay this time double since we would need to revert our changes then when the change from Blender arrives.This time is better invested elsewhere.

I undestand your needs. It is this crazy implemented that even the Blender devs realizes that it needs improvement. But i have to take care of the big picture of Bforartists. And this means i cannot give green light for such changes, sorry.

What we can do though is changes that doesn't modify the current brush system. But sits on top of it. We could for example expose the sub brush set below the header. Maybe even an addon already exists for that. But then again, most brushes does have just one pencil anyways. So you first would need to document what exact needs you have. One by one. Rewriting the brush system as a whole is no task that we can manage here. And nothing that i will accept as a commit.

![solutions](https://user-images.githubusercontent.com/13919702/104280591-7214b580-54ac-11eb-956d-56dd869fecc0.jpg)

**Comment by Draise14** [comment]
Hopefully there is no rewriting at all but added duplicate code and find the way to just set the duplicate default to the tools shelf button. The idea is find a way to add easily.. not change nor rewrite. 

Just let us hit the wall if there is one. Learn something. It's not a rewrite nor redesign. 

The exact need is simple. 7 more brushes in the toolshelf for one click mouse activation in weight paint, vertex paint, and textures paint. 

Duplicate brush code. See how to have it assign to a different brush setting. That's it. Same code. Same library or duplicate library to new brush duplicate.

**Comment by iyadahmed** [comment]
The task is not impossible but indeed tricky, here's the info we have now, summarized:

1. Tools are defined in `release\scripts\startup\bl_ui\space_toolsystem_toolbar.py`
2. Many brushes are auto generated from the enums defined in `source\blender\makesrna\intern\rna_brush.c`
and accessible using `bpy.types.Brush.bl_rna.properties[attr].enum_items_static`
where attr can be `weight_tool` for weight brushes for example

the auto genration is done using generate_from_enum_ex in space_toolsystem_toolbar.py

3. the actual drawing happens in `release\scripts\startup\bl_ui\space_toolsystem_common.py`
it also automatically reads defined tools by using ToolSelectPanelHelper.__subclasses__()

where ToolSelectPanelHelper is the base class, which defined tools inherit from.

it is easy duplicate one of the already existing tools (succeeded in that), but not change it to use Subtract brush for example,
because only the Draw brush is defined in the enums, and not its "sub brushes", I will continue to try at this

**Comment by Draise14** [comment]
Yeah. I was guessing that. The enum list would be dynamic to the brush type and not set unless... By user? Where is that stored?

 If a duplicate enum list was separate or new with another copy also, maybe you could have two libraries exposed to two brushes?

**Comment by iyadahmed** [comment]
> Yeah. I was guessing that. The enum list would be dynamic to the brush type and not set unless... By user? Where is that stored?
> 
> 
> 
>  If a duplicate enum list was separate or new with another copy also, maybe you could have two libraries exposed to two brushes?

The enum lists are unfortunately static and defined in rna_brush.c, we can access brushes directly using bpy.data.brushes, which lists all brushes, but toolshelf system relies on the enums, which also the brush system uses
![image](https://user-images.githubusercontent.com/25552173/107108608-e50e1380-6841-11eb-9e8b-2dfb309a4456.jpeg)

**Comment by Draise14** [comment]
# Conclusion
You can duplicate the brush, but just makes a new brush
You cannot define which enum the duplicate brush goes to by default, as that is stored in the blend file practically.
The brushes are dynamic and setup inside the blend file, meaning that you'd need a whole new brush library instead of filtering current libraries. 

So.. duplicating a brush and library would mean.. practically making a new brush from scratch to mimic a new brush library and their brush data blocks. 

TLDR: you'd have to make a whole new brush with a whole new library to "expose" different brush groups to the surface level - but the code is so convoluted to be dynamic, it's not practical to make new brushes. 

_This means... I will probably chew things over to see what should be exposed to a top level or to header for another kind of proposal, and think about maybe better pie menus for these tools to just.. not have to use that drop down menu if at all and just paint._
